### ISSUE TITLE
linter: Add a type-aware rule to detect usage of never values

### ISSUE THREAD

**Issue by mathcovax** [oxc-project/oxc#26766]
It would be useful to add a TypeScript type-aware rule that reports when a value whose inferred type has been narrowed to `never` is still being used.

Since `never` is assignable to every type in TypeScript, this remains valid:

```ts
consume(value); // `value` is `never`
```

This can hide dead code or logic mistakes, especially when narrowing happens through predicates and functional abstractions.

## Example

A concrete case can happen inside a pipeline:

```ts
const result = pipe(
	right("success"),
	when(
		isLeft,
		(value) => {
			value satisfies never;

			consume(value); // accepted by TypeScript
			return value;
		},
	),
);
```

Here, `right("success")` produces a value already known to be a `Right`.

The `isLeft` predicate narrows its input using:

```ts
Extract<Input, Left>
```

In this context:

```ts
Extract<Right, Left>
```

resolves to `never`.

TypeScript therefore correctly infers that `value` is `never`, which can be verified with:

```ts
value satisfies never;
```

However, `consume(value)` is still accepted because `never` is assignable to any type.

The callback can never be executed at runtime, so this is effectively dead code.

[TypeScript Playground reproduction](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAMhBmwoF4oG8BQUqkgLigCIAbBYQgbiyggCdaB7WggUXqaoF8qNdoAlAJYBzABZJUmbHwKFaI8ZWoA3AIbEArhAIBnYPIB2wrjwDGDA3qjyxEqAAo1m7VD2HhASgJDbKAHwOUjjgLnIK5AA0KupaUZwePPAaBqbAghZQgjpwiAA8ACp+jjEu+V5QTlqZOlAsAB76qqkFEbBkAUG0EMAatAbBkAzwFSUoyKiEDABGAFYQqYTU2ABkyyPOUACE41AGGsTES1CrRHyEmf2VEEcnVwB0fGMTpIhKnBgYSSlpGQDuohADLkAJIGMAaYCtAByqnYvwgABMaA1AQiaqDwZCoAB5CGYorUMBdBGCUyqYAueyCMEQggYiEefwXTHVKAwuGIqLSWiqSzwJgAWwIVJpwAI7MY8IRjOQAVxwExUUZnW6vX6IsxdNF5XlLIAPrU6qZNAiICDRdDYZLEQFZUciYjSeSIBqGUdsAB+HA8vmC13ADzuqAEaks1Q1erGjSm82KtlWhhSvxcD5fVLpfpgQSQWMQ1q6iEE7Ch2lQemQ6iNX20IUOEtisvapkFis6vEQ9DULo9Pre3k6fk1-0JDDvT7JdMZcyWDQCl1XXT6ameTvYac6BikO7EBjCYrOEdj9dILo6fZ2LOQexd8L2Qhn0ymCA6HSEDxcqD-QHX7DF7JkD9sH3LQZQ6IMrlccksngQRn12CBlDoKhf1-ddZ3nEoEigAB6bCoCaJ8wApJEEMBHBRAYDQxCgAADK4aNZGiDFI2gaKDbs1XWLRkN-TgP3fDARwwddNwgbdd3sU9zxHIA)

## Why `no-unnecessary-condition` is not enough

`typescript/no-unnecessary-condition` works well when the condition is directly visible to the linter.

In this example, however, the control flow is encapsulated inside a function such as `when`.

The linter does not need, and probably should not need, to understand the semantics of `when`, `pipe`, `match`, or similar abstractions.

The type checker already has the relevant information: **the value passed to the callback is `never`**.

## Proposed behavior

A rule such as:

```txt
typescript/no-never-value
```

could report usages of expressions whose resolved type is `never`:

```ts
consume(value);
//      ^ error: value has type `never`

const result = value;
//             ^ error

array.push(value);
//         ^ error
```

The goal would not be to forbid the `never` type itself.

Intentional uses such as exhaustiveness helpers should remain valid:

```ts
function assertNever(value: never): never {
	throw new Error("Unexpected value");
}
```

## Motivation

This rule could help detect:

* dead code hidden behind functional abstractions;
* predicates applied to types they can no longer match;
* branches made impossible after refactoring;
* narrowing mistakes that are difficult to notice;
* cases that `no-unnecessary-condition` cannot easily analyze.

The signal is relatively simple: a value that TypeScript already considers impossible (`never`) is still being consumed as a regular value.

## AI assistance disclosure

This issue was drafted with the assistance of AI, then manually reviewed and edited before submission.

The goal of using AI here was only to help structure and phrase the proposal. The technical reasoning, example, and intent were reviewed by the author, and this issue is not intended to contribute to low-quality AI-generated issue spam or "AI slop".
